src/tscore/ThrottledMessage.cc

```diff
@@ -0,0 +1,96 @@
+/** @file
+
+  ThrottledMessage declaration.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "tscore/ThrottledMessage.h"
+
+ThrottledMessage::ThrottledMessage(std::chrono::seconds seconds_between)
+  : _last_log_time{TimePoint{}}, _seconds_between{seconds_between}
+{
+}
+
+ThrottledMessage::ThrottledMessage(ThrottledMessage &&other)
+  : _last_log_time{other._last_log_time.load()},
+    _seconds_between{other._seconds_between},
+    _count_since_last_log{other._count_since_last_log}
+{
+}
+
+void
+ThrottledMessage::ThrottleStatus(const char *fmt, ...)
+{
+  uint32_t count_since_last_log = 0;
+  if (!should_emit_log(count_since_last_log)) {
+    return;
+  }
+  Status("Skipped the following message %d times.", count_since_last_log);
+
+  va_list args;
+  va_start(args, fmt);
+  StatusV(fmt, args);
+  va_end(args);
+}
+
+void
+ThrottledMessage::ThrottleNote(const char *fmt, ...)
+{
+  uint32_t count_since_last_log = 0;
+  if (!should_emit_log(count_since_last_log)) {
+    return;
+  }
+  Note("Skipped the following message %d times.", count_since_last_log);
+
+  va_list args;
+  va_start(args, fmt);
+  NoteV(fmt, args);
+  va_end(args);
+}
+
+void
+ThrottledMessage::ThrottleWarning(const char *fmt, ...)
+{
+  uint32_t count_since_last_log = 0;
+  if (!should_emit_log(count_since_last_log)) {
+    return;
+  }
+  Warning("Skipped the following message %d times.", count_since_last_log);
+
+  va_list args;
+  va_start(args, fmt);
+  WarningV(fmt, args);
+  va_end(args);
+}
+
+void
+ThrottledMessage::ThrottleError(const char *fmt, ...)
```

Comment:
These methods look *very* similar. Is there no way to combine them?

Reply:
Yeah, I mention this frustration in the description. In similar cases in the past I've had a helper function that takes a std::function, but Status, Note, etc., are macros.  I'm not sure how to easily consolidate. I suppose I can add another macro...

Reply:
I was able to clean this up by calling the diags->error function directly, rather than through the logging macros.